Palette settings

idotobi opened the pull request:
Add UI to change palette inside xournalpp Settings.

This would properly close #1812.

idotobi commented:
I finally managed to get the test setup for the "integration-tests" working :tada: . Sorry for the long wait :bow: 

Additional to the automated tests I manually tested the following on Linux (Ubuntu 22.04.4 LTS):
1. remove existing `~/.config/xournalpp`
2. checkout `master` and build and install
3. start xournalpp to see if it is working
4. checkout `palette-settings` branch and build and install
5. start xournalpp and change the palette in the settings
6. close xournalpp and start again to see if change got persisted
7. customize toolbar (`View` -> `Toolbars` -> `Customize`) and drag several color items in between the dialog and the toolbar
8. add a palette file to `~/.config/xournalpp/palettes` (the directory should already exist as long as you already started xournalpp)
9. go to settings and check whether the new palette is listed

Additional to code review I think it would be good if the same manual test as above could be repeated for

- [ ] Mac
- [ ] Windows

I only have Linux, so it would be great if someone else could try it out.

idotobi commented:
A Note for the other maintainers:

- I implemented an `integration-test` setup that allows to test code that relies on gtk stuff. It's far from perfect but at least allows for some basic smoke tests like "... does not throw"/"... does not cause segmentation fault". Please judge for yourself if the additional complexity is worth it.
- I reordered the commits a bit to make them easier to review, before merge I would further squash them. However, I though it might be easier to review with still some commit granularity.

rolandlo commented:
Thanks @idotobi for the update. Currently it is crashing on MacOS (when run via jhbuild):

```shell
[bt] #0 1   xournalpp                           0x00000001002c3313 _ZL12crashHandleri + 2515
[bt] #1 2   libsystem_platform.dylib            0x00007fff6d37f5fd _sigtramp + 29
[bt] #2 3   ???                                 0x00007f818ec9b3c0 0x0 + 140194423092160
[bt] #3 4   xournalpp                           0x00000001002ad861 _ZN7Control23loadPaletteFromSettingsEv + 305
[bt] #4 5   xournalpp                           0x00000001002ad17c _ZN7ControlC2EP13_GApplicationP15GladeSearchpathb + 684
[bt] #5 6   xournalpp                           0x00000001002daf64 _ZN12_GLOBAL__N_110on_startupEP13_GApplicationPNS_18XournalMainPrivateE + 3076
[bt] #6 7   libgobject-2.0.0.dylib              0x0000000102638284 g_closure_invoke + 196
[bt] #7 8   libgobject-2.0.0.dylib              0x000000010264f700 signal_emit_unlocked_R + 1632
[bt] #8 9   libgobject-2.0.0.dylib              0x000000010265051d g_signal_emit_valist + 2157
[bt] #9 10  libgobject-2.0.0.dylib              0x0000000102650c72 g_signal_emit + 130
[bt] #10 11  libgio-2.0.0.dylib                  0x0000000102424bb8 g_application_register + 280
[bt] #11 12  libgio-2.0.0.dylib                  0x0000000102427247 g_application_real_local_command_line + 1559
[bt] #12 13  libgio-2.0.0.dylib                  0x00000001024252b2 g_application_run + 450
[bt] #13 14  xournalpp                           0x00000001002d9982 _ZN11XournalMain3runEiPPc + 1810
[bt] #14 15  libdyld.dylib                       0x00007fff6d186cc9 start + 1
```

Did I miss a step?

idotobi commented on src/core/control/Control.cpp:
```diff
@@ -2373,3 +2378,22 @@ auto Control::getPageBackgroundChangeController() const -> PageBackgroundChangeC
 
 auto Control::getPluginController() const -> PluginController* { return this->pluginController; }
+
+auto Control::getPalette() const -> const Palette& { return *(this->palette); }
```
- [ ] Is this the correct way to implement this? I'm not to sure since C++ is not my day-to-day language.

The intention is to simply return the reference to the one "global" `Palette` instance that's managed by `Control`. Basically a singleton since `Palette` is also instantiated only once.

idotobi commented:
@rolandlo: No, I think that might be a valid finding. I just pushed a potential fix.

rolandlo commented:
Thanks @idotobi . The segfault doesn't appear anymore. There seems to be an issue with the path though:
![grafik](https://github.com/xournalpp/xournalpp/assets/40485433/e9ab56f8-c640-4009-a192-f40db2e1d234)

The file is located at `/Users/xournal-dev/gtk/inst/share/xournalpp/palettes/xournal_classic.gpl` in reality.

rolandlo commented:
After copying the file to the place Xournal++ currently expects it, I see the palette in the settings. The location of the palette is hardly legible though in light mode.

![grafik](https://github.com/xournalpp/xournalpp/assets/40485433/c8e299a3-cb87-4ea1-b8e6-bc4b71e79e8f)

In dark mode it's OK. 

![grafik](https://github.com/xournalpp/xournalpp/assets/40485433/325fdc61-778f-4115-8d43-c251133199da)

rolandlo commented:
I have copied the other palette as well:
![grafik](https://github.com/xournalpp/xournalpp/assets/40485433/6b2c4769-856d-4272-978c-fc4bbd67bd89)

Please don't call it "Xournal new palette" though. It has nothing to do with Xournal. We add to the confusion about Xournal/Xournal++ by using the name Xournal when Xournal++ is meant.

rolandlo commented:
There are some issues with toolbar customization regarding palette colors: Removing some colors from the top toolbar, dragging colors from the dialog to the bottom toolbar and then dragging the added colors to the top toolbar results in wrong color names and a crash when moving too many colors up there. It might be that those issues are in master as well (I will have to check) and will be fixed with @bhennion's branch allowing multiple instances of the same tool in the same toolbar.

Another thin I noticed: On Azure Pipelines the integration tests are also called "Run tests" which is confusing.
![grafik](https://github.com/xournalpp/xournalpp/assets/40485433/f39e74fe-8fd2-4e69-ab5d-9fcd748f7e62)

rolandlo commented:
The palette set in the settings is remembered correctly on restart and palettes created in the `.config/xournalpp/palettes` folder show up correctly. The UX feels pretty good.

Febbe commented on src/core/control/XournalMain.h:
```diff
@@ -12,8 +12,11 @@
 
 
+#include <glib.h>
 
-
+#include "gui/GladeSearchpath.h"
```
``` suggestion
class GladeSearchpath;
```

Febbe commented on src/core/control/XournalMain.h:
```diff
@@ -12,8 +12,11 @@
 auto run(int argc, char** argv) -> int;
-}
+
+void initResourcePath(GladeSearchpath* gladePath, const gchar* relativePathAndFile, bool failIfNotFound = true);
```
```suggestion
void initResourcePath(GladeSearchpath* gladePath, std::string_view relativePathAndFile, bool failIfNotFound = true);
```
Is this function required by other compilation units?

Febbe commented on src/core/control/settings/Settings.h:
```diff
@@ -569,7 +570,8 @@ class Settings {
     void setStabilizerPreprocessor(StrokeStabilizer::Preprocessor preprocessor);
 
-    const Palette& getColorPalette();
+    const std::optional<fs::path> getColorPaletteSetting();
```
```suggestion
    auto getColorPaletteSetting() -> fs::path const&;
```
Returning a const value may force a double copy, the copy won't be const anymore.
Returning an optional to an implicitly on the heap allocated value is over verbose, since an empty path would be the same.

Febbe commented on src/core/control/settings/Settings.h:
```diff
@@ -569,7 +570,8 @@ class Settings {
 
-    const Palette& getColorPalette();
+    const std::optional<fs::path> getColorPaletteSetting();
+    void setColorPaletteSetting(const std::optional<fs::path>& palettePath);
```
```suggestion
    void setColorPaletteSetting(fs::path palettePath);
```

Febbe commented on src/core/control/settings/Settings.h:
```diff
@@ -1151,12 +1153,7 @@ class Settings {
-     */
-    std::unique_ptr<Palette> palette;
-
+    std::optional<fs::path> colorPaletteSetting;
```
```suggestion
    fs::path colorPaletteSetting;
```

Febbe commented on src/core/gui/dialog/SettingsDialog.cpp:
```diff
@@ -16,16 +16,16 @@
-#include "util/i18n.h"                           // for _
-#include "util/raii/CairoWrappers.h"             // for CairoSurfaceSPtr
-#include "util/safe_casts.h"                     // for round_cast
+#include "gui/toolbarMenubar/ToolMenuHandler.h"  // for ToolMenuHandler
+#include "gui/toolbarMenubar/icon/ColorIcon.h"
+#include "gui/toolbarMenubar/model/ColorPalette.h"  // for Palette
+#include "gui/widgets/ZoomCallib.h"  // for zoomcallib_new, zoom...#include "model/PageType.h"                         // for PageType
+#include "model/PageType.h"   // for PageType
+#include "util/Color.h"       // for GdkRGBA_to_argb, rgb...
+#include "util/PathUtil.h"    // for fromGFile, toGFilename
+#include "util/Util.h"        // for systemWithMessage#include "util/gtk4_helper.h"                       //
+#include "util/i18n.h"        // for _#include "util/raii/CairoWrappers.h"                // for CairoSurfaceSPtr
+#include "util/safe_casts.h"  // for round_cast
```
Many includes moved into the comment of the line above.

idotobi replied:
:facepalm: Don't know how that happend. But thank you very much for catching it!

Febbe commented on src/core/gui/toolbarMenubar/model/ColorPalette.cpp:
```diff
@@ -127,7 +133,7 @@ void Palette::create_default(fs::path filepath) {
 auto Header::getAttribute() const -> std::string { return this->attribute; };
 auto Header::getValue() const -> std::string { return this->value; };
-
+auto Palette::getFilePath() const -> fs::path { return this->filepath; };
```
```suggestion
auto Palette::getFilePath() const -> fs::path const& { return this->filepath; };
```

Febbe commented on src/core/gui/toolbarMenubar/model/ColorPalette.h:
```diff
@@ -127,6 +127,10 @@ struct Palette {
 
+    fs::path getFilePath() const;
+
+    std::basic_string<char> getHeader(const std::string& attr);
```
```suggestion
    std::string getHeader(const std::string& attr);
```

Febbe commented on src/core/gui/dialog/SettingsDialogPaletteTab.cpp:
```diff
@@ -0,0 +1,193 @@
+auto SettingsDialogPaletteTab::getSelectedPalette() -> std::optional<fs::path> {
+    GtkListBoxRow* selected_listbox_row = gtk_list_box_get_selected_row(paletteListBox);
+    if (allPaletteFilePaths.empty()) {
+        return std::nullopt;
+    }
+    if (!G_IS_OBJECT(selected_listbox_row)) {
+        return std::nullopt;
```
```suggestion
        return {};
    }
    if (!G_IS_OBJECT(selected_listbox_row)) {
        return {};
```
Unfortunately, std::optional does not support an overload, to just use the state of the type in it.

idotobi commented:
> Thanks @idotobi . The segfault doesn't appear anymore. There seems to be an issue with the path though: ![grafik](https://private-user-images.githubusercontent.com/40485433/318327759-e9ab56f8-c640-4009-a192-f40db2e1d234.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIwODY4NDYsIm5iZiI6MTcxMjA4NjU0NiwicGF0aCI6Ii80MDQ4NTQzMy8zMTgzMjc3NTktZTlhYjU2ZjgtYzY0MC00MDA5LWExOTItZjQwZGIyZTFkMjM0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDAyVDE5MzU0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBmYjYwOGMxNTBiNTRlZmNkZjdjMGE2NTYzMGVhMjQxNjE5YWQyYjc1YmM1MWEyZjJmNzQzY2QzM2ZhZTM2YWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.5j-XuH7c2s60jHibZaQQ63yLGM3SiTO-e0hT2lQOuA8)
> 
> The file is located at `/Users/xournal-dev/gtk/inst/share/xournalpp/palettes/xournal_classic.gpl` in reality.

Interesting: Seems like I used some assumption that's only valid on Linux when trying to find the palette file.
`/Users/xournal-dev/gtk/inst/share/xournalpp/palettes/xournal_classic.gpl` sounds like it should be the correct path. Can you give me some example files/directories listed under `/Users/xournal-dev/gtk/inst/share/xournalpp/`. This probably makes it way faster for me to find the correct method.

idotobi commented:
> The location of the palette is hardly legible though in light mode.

That's true, I'll try to set a darker gray for light mode.

idotobi commented:
> Please don't call it "Xournal new palette" though. It has nothing to do with Xournal. We add to the confusion about Xournal/Xournal++ by using the name Xournal when Xournal++ is meant.

Good point. Will do so.

idotobi commented:
> There are some issues with toolbar customization regarding palette colors: Removing some colors from the top toolbar, dragging colors from the dialog to the bottom toolbar and then dragging the added colors to the top toolbar results in wrong color names and a crash when moving too many colors up there. It might be that those issues are in master as well (I will have to check) and will be fixed with @bhennion's branch allowing multiple instances of the same tool in the same toolbar.

Ah, damn. I was hoping that I fixed this :see_no_evil: . I had the exact same problem on my machine at some point, but fixed it. But I guess it's not fully fixed for all OSs. Still would be good if you can check if you can reproduce it on master on MacOs, just to make sure whether I introduced this bug. You can also check whether my last commit somehow fixed it. I would be surprised if, but maybe I got lucky :sweat_smile: 


> Another thin I noticed: On Azure Pipelines the integration tests are also called "Run tests" which is confusing.

Yeah, I overlooked this when I did my review. It should be something like "Run gtk-integration-tests".

rolandlo commented:
> > Thanks @idotobi . The segfault doesn't appear anymore. There seems to be an issue with the path though: ![grafik](https://private-user-images.githubusercontent.com/40485433/318327759-e9ab56f8-c640-4009-a192-f40db2e1d234.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIwODY4NDYsIm5iZiI6MTcxMjA4NjU0NiwicGF0aCI6Ii80MDQ4NTQzMy8zMTgzMjc3NTktZTlhYjU2ZjgtYzY0MC00MDA5LWExOTItZjQwZGIyZTFkMjM0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDAyVDE5MzU0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBmYjYwOGMxNTBiNTRlZmNkZjdjMGE2NTYzMGVhMjQxNjE5YWQyYjc1YmM1MWEyZjJmNzQzY2QzM2ZhZTM2YWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.5j-XuH7c2s60jHibZaQQ63yLGM3SiTO-e0hT2lQOuA8)
> > The file is located at `/Users/xournal-dev/gtk/inst/share/xournalpp/palettes/xournal_classic.gpl` in reality.
> 
> Interesting: Seems like I used some assumption that's only valid on Linux when trying to find the palette file. `/Users/xournal-dev/gtk/inst/share/xournalpp/palettes/xournal_classic.gpl` sounds like it should be the correct path. Can you give me some example files/directories listed under `/Users/xournal-dev/gtk/inst/share/xournalpp/`. This probably makes it way faster for me to find the correct method.

In `/Users/xournal-dev/gtk/inst/share/xournalpp/` along with `palettes`  we have `plugins`, `resources`, `ui`.

rolandlo commented:
The palettes folder is not copied over to the .app bundle. The folder `Xournal++.app/Contents/Resources/palettes` is missing. In `Xournal++.app/Contents/Resources/` there are currently only `lib/`, `plugins/`, `share/` and `ui/`. The copying should be handled in `mac-setup/xournalpp.bundle`

idotobi commented:
@rolandlo 
I updated the PR with the following:

- [x] improve readability of palette path in both light and dark mode (this change also propagates the darkMode css class to dialog windows now :sunglasses:. If you disagree with the values it's now easy to fix them in css.
- [x] rename gtk-integration-tests in pipeline
- [x] rename Palettes to "Xournal Palette" (old classic) and "Xournal++ Palette" for the new one. I guess that's both historically correct and should make the distinction quite clear. We could also call it "Xournal++ Default Palette", but that's probably to verbose :thinking: 
- [x] adding the palette files to the mac bundle (or at least I think I did if I understood the setup correctly :crossed_fingers: )
- [x] I also had to remove the `SettingsDialogTest.cpp` :disappointed:, since the setup broke after the last rebase and I couldn't get it running within 3h. Given that I wrote it and couldn't properly debug it, I'd argue it's not a stable enough setup to have in master yet, and it was only smoke tests anyways.

Since I rebased, it could very well be that the new Toolbar Customization also fixed the Segfault. At least on my system (Ubuntu) I didn't get a segfault, but since I didn't get one the last time that doesn't say too much :sweat_smile:

rolandlo commented:
@idotobi 
Currently the `DEV_DEFAULT_PALETTE_FILE "xournal_classic.gpl"` doesn't match the filename `xournal.gpl`. Moreover, after fixing that, the palettes don't show up in the palette settings even though the correct directories are listed and the palette files exist in the listed directories. So could you please check what you pushed?

idotobi commented:
Odd, I tested it quite a bit after the merge and didn't have any issues :thinking:. 
Still sorry for the inconvenience.

I fixed the `DEV_DEFAULT_PALETTE_FILE now, but I cannot reproduce the issue

> Moreover, after fixing that, the palettes don't show up in the palette settings even though the correct directories are listed and the palette files exist in the listed directories.

I just repeated the same test again as I did after the rebase, namely:

> 1. remove existing `~/.config/xournalpp`
> 2. checkout `master` and build and install
> 3. start xournalpp to see if it is working
> 4. checkout `palette-settings` branch and build and install
> 5. start xournalpp and change the palette in the settings
> 6. close xournalpp and start again to see if change got persisted
> 7. customize toolbar (`View` -> `Toolbars` -> `Customize`) and drag several color items in between the dialog and the toolbar
> 8. add a palette file to `~/.config/xournalpp/palettes` (the directory should already exist as long as you already started xournalpp)
> 9. go to settings and check whether the new palette is listed

For me they show up as expected:

![image](https://github.com/xournalpp/xournalpp/assets/16611056/43b90d56-9295-4099-a9b3-66c97fde0a6c)

Could it be that my attempt to get it working for Mac broke something (https://github.com/xournalpp/xournalpp/pull/3767/commits/0ff6b50f392b59a4b1fb34ca14425bf91bdb8f32)?
I assume that the palettes in the "built-in" path were not found, but the user palettes worked?
Where do the links in in the Palette Settings link to? Do you see any files there?

Maybe it has something to do with the [`getDataPath`](https://github.com/idotobi/xournalpp/blob/4836d638cbed2cd83491994841953494fba9297a/src/util/PathUtil.cpp#L306-L326) which in turn is used by [`getPalettePath`](https://github.com/idotobi/xournalpp/blob/4836d638cbed2cd83491994841953494fba9297a/src/util/PathUtil.cpp#L338) :thinking:

rolandlo commented:
Indeed after removing my old `settings.xml` the palettes show up again. 
Note though that users won't just remove their `settings.xml` when we merge the PR. We need to take that into account.

rolandlo commented:
Minor issue: When I switch from dark mode to light mode in my OS, the highlighting of the palette paths doesn't update automatically (in contrast to the rest of the GUI).

Bigger issue: I still managed to get a signal 6 crash while moving colors between the toolbars:
```shell
Try to get a better stacktrace...
[bt] #0 inst/bin/xournalpp(+0x46d0ed) [0x5bb9b4f140ed]
crashHandler(int) bei /home/roland/dev/Reviews/xournalpp/src/core/control/CrashHandlerUnix.h:156
[bt] #1 /lib/x86_64-linux-gnu/libc.so.6(+0x45320) [0x75e8d1445320]
__restore_rt bei libc_sigaction.c:?
[bt] #2 /lib/x86_64-linux-gnu/libc.so.6(pthread_kill+0x11c) [0x75e8d149eb1c]
__pthread_kill_implementation bei ./nptl/./nptl/pthread_kill.c:44 (discriminator 1)
[bt] #3 /lib/x86_64-linux-gnu/libc.so.6(gsignal+0x1e) [0x75e8d144526e]
__GI_raise bei ./signal/../sysdeps/posix/raise.c:27
[bt] #4 /lib/x86_64-linux-gnu/libc.so.6(abort+0xdf) [0x75e8d14288ff]
__GI_abort bei ./stdlib/./stdlib/abort.c:81
[bt] #5 /lib/x86_64-linux-gnu/libstdc++.so.6(+0xa5ffe) [0x75e8d18a5ffe]
std::unexpected() bei ??:?
[bt] #6 /lib/x86_64-linux-gnu/libstdc++.so.6(+0xbae9c) [0x75e8d18bae9c]
std::rethrow_exception(std::__exception_ptr::exception_ptr) bei ??:?
[bt] #7 /lib/x86_64-linux-gnu/libstdc++.so.6(_ZSt10unexpectedv+0) [0x75e8d18a5a49]
std::unexpected() bei ??:?
[bt] #8 /lib/x86_64-linux-gnu/libstdc++.so.6(+0xbb128) [0x75e8d18bb128]
__cxa_throw bei ??:?
[bt] #9 /lib/x86_64-linux-gnu/libstdc++.so.6(+0xa5abb) [0x75e8d18a5abb]
std::unexpected() bei ??:?
[bt] #10 inst/bin/xournalpp(_ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE12_M_constructIPcEEvT_S7_St20forward_iterator_tag+0x77) [0x5bb9b4f537f5]
void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct<char*>(char*, char*, std::forward_iterator_tag) bei /usr/include/c++/13/bits/basic_string.tcc:229
[bt] #11 inst/bin/xournalpp(_ZN10NamedColorC1ERKS_+0x3d) [0x5bb9b4fdde95]
NamedColor::NamedColor(NamedColor const&) bei /home/roland/dev/Reviews/xournalpp/src/util/include/util/NamedColor.h:26 (discriminator 1)
[bt] #12 inst/bin/xournalpp(_ZSt11make_uniqueI13ColorToolItemJRK10NamedColorEENSt8__detail9_MakeUniqIT_E15__single_objectEDpOT0_+0x58) [0x5bb9b4fddf1d]
std::__detail::_MakeUniq<ColorToolItem>::__single_object std::make_unique<ColorToolItem, NamedColor const&>(NamedColor const&) bei /usr/include/c++/13/bits/unique_ptr.h:1070 (discriminator 2)
[bt] #13 inst/bin/xournalpp(_ZN14ToolbarAdapter25toolbarDragDataReceivedCbEP11_GtkToolbarP15_GdkDragContextiiP17_GtkSelectionDatajjPS_+0x2f4) [0x5bb9b51da294]
ToolbarAdapter::toolbarDragDataReceivedCb(_GtkToolbar*, _GdkDragContext*, int, int, _GtkSelectionData*, unsigned int, unsigned int, ToolbarAdapter*) bei /home/roland/dev/Reviews/xournalpp/src/core/gui/dialog/toolbarCustomize/ToolbarAdapter.cpp:281
[bt] #14 /lib/x86_64-linux-gnu/libgtk-3.so.0(+0x9b276) [0x75e8d229b276]
?? ??:0
[bt] #15 /lib/x86_64-linux-gnu/libgobject-2.0.so.0(+0x356bd) [0x75e8d1f096bd]
g_signal_handlers_destroy bei ??:?
[bt] #16 /lib/x86_64-linux-gnu/libgobject-2.0.so.0(g_signal_emit_by_name+0x1f8) [0x75e8d1f09a98]
g_signal_emit_by_name bei ??:?
```

rolandlo commented:
Our path handling is messed up what concerns `getDataPath()`
On Linux: `/usr/share/xournalpp`
On MacOS: `$HOME/gtk/inst` when running with jhbuild and `Xournal++.app/Contents/Resources` when running as app bundle

So `getPalettePath()`:

On Linux: `/usr/share/xournalpp/palettes` (good)
On MacOS: `$HOME/gtk/inst/palettes` (bad; when running with jhbuild), `Xournal++.app/Contents/Resources/palettes` (good; when running as app bundle)

On MacOS `getDataPath` should be fixed (for jhbuild runs) but then `getLocalePath` should be fixed as well. I'll make a suggestion for a fix.

rolandlo commented:
> Our path handling is messed up what concerns `getDataPath()` On Linux: `/usr/share/xournalpp` On MacOS: `$HOME/gtk/inst` when running with jhbuild and `Xournal++.app/Contents/Resources` when running as app bundle
> 
> So `getPalettePath()`:
> 
> On Linux: `/usr/share/xournalpp/palettes` (good) On MacOS: `$HOME/gtk/inst/palettes` (bad; when running with jhbuild), `Xournal++.app/Contents/Resources/palettes` (good; when running as app bundle)
> 
> On MacOS `getDataPath` should be fixed (for jhbuild runs) but then `getLocalePath` should be fixed as well. I'll make a suggestion for a fix.

This works:
```patch
diff --git a/src/util/PathUtil.cpp b/src/util/PathUtil.cpp
index e42a8fe33..305219993 100644
--- a/src/util/PathUtil.cpp
+++ b/src/util/PathUtil.cpp
@@ -316,6 +316,9 @@ auto Util::getDataPath() -> fs::path {
     fs::path p = Stacktrace::getExePath().parent_path();
     if (fs::exists(p / "Resources")) {
         p = p / "Resources";
+    } else {
+        p = PACKAGE_DATA_DIR; 
+        p /= PROJECT_NAME;
     }
     return p;
 #else
@@ -326,13 +329,14 @@ auto Util::getDataPath() -> fs::path {
 }
 
 auto Util::getLocalePath() -> fs::path {
-#ifdef _WIN32
-    return getDataPath() / ".." / "locale";
-#elif defined(__APPLE__)
-    return getDataPath() / "share" / "locale";
-#else
+#ifdef __APPLE__
+    fs::path p = Stacktrace::getExePath().parent_path();
+    if (fs::exists(p / "Resources")) {
+        return p / "Resources" / "share" / "locale";
+    }
+#endif
+
     return getDataPath() / ".." / "locale";
-#endif
 }
 
 auto Util::getPalettePath() -> fs::path { return getDataPath() / "palettes"; }```

idotobi commented:
> Indeed after removing my old `settings.xml` the palettes show up again.
> Note though that users won't just remove their `settings.xml` when we merge the PR. We need to take that into account.

I assume this problem was not just caused by the path issue on Mac :see_no_evil: 
If not, could you send me the two conflicting `settings.xml` files (before and after the checkout of the branch)?
Since it didn't happen in my test steps I assume it's something I cannot easily reproduce on Linux, as I designed the test steps particularly to check if something like this happening :facepalm:

idotobi commented:
> Minor issue: When I switch from dark mode to light mode in my OS, the highlighting of the palette paths doesn't update automatically (in contrast to the rest of the GUI).

Seems like the implementation in https://github.com/idotobi/xournalpp/commit/aeda91d20dcc075c89a67de457253e6d1d8bc16d doesn't update the css on the fly. When changing through the settings menu it was not noticeable as the settings only apply after saving. I'll have a look at it.

idotobi commented:
> Bigger issue: I still managed to get a signal 6 crash while moving colors between the toolbars

:see_no_evil: Ah, alright. Then I'll try my best to analyze it, but I assume that will take some time.
Still, I would definitely not want to merge it with an issue like that still around. Thanks for checking on it :pray:

idotobi commented:
The path fix is added now.

rolandlo commented:
> > Indeed after removing my old `settings.xml` the palettes show up again.
> > Note though that users won't just remove their `settings.xml` when we merge the PR. We need to take that into account.
> 
> I assume this problem was not just caused by the path issue on Mac 🙈 If not, could you send me the two conflicting `settings.xml` files (before and after the checkout of the branch)? Since it didn't happen in my test steps I assume it's something I cannot easily reproduce on Linux, as I designed the test steps particularly to check if something like this is happening 😓

It has nothing to do with MacOS, since the changes there can't affect Linux. Here is my `settings.xml` file (just remove the .txt ending). 
[settings.xml](https://github.com/xournalpp/xournalpp/files/15129361/settings.xml.txt)

idotobi commented:
@rolandlo : Regarding the `settings.xml` problem:
So I compared your `settings.xml` with the ones from my setup when on `main` and when on this branch.
 - **yours vs. my main:** only minor settings differences. Nothing suspicious.
 - **yours vs. my branch:** only additional `<property name="colorPalette" value="/home/idotobi/Documents/private/projects/xournalpp/xournalpp/cmake-build-debug/install/share/xournalpp/palettes/xournalpp.gpl"/>`

Going through the code this cannot really be the cause.
When loading the settings.xml we only overwrite the local palette variable when the settings property exists and is not empty (see [Settings.cpp#L680-L685](https://github.com/idotobi/xournalpp/blob/13658feef3617a0382d420503a5f43145f5b408f/src/core/control/settings/Settings.cpp#L680-L685)).
The default (which is set in `loadDefault` also doesn't depend on the settings (see [Settings.cpp#L243-L244](https://github.com/idotobi/xournalpp/blob/13658feef3617a0382d420503a5f43145f5b408f/src/core/control/settings/Settings.cpp#L243-L244)).
Hence, given that your settings file does not contain the `colorPalette` property, the default will not be overwritten and consequently the default `colorPalette` property should be written the next time the settings are saved. Thus, this is probably not the cause of the issue. 

So lets look at it from a different direction.
If I understood you correctly then you didn't see any palettes in the palette settings. So you probably see something like the following, right?

![image](https://github.com/xournalpp/xournalpp/assets/16611056/2c38deba-310e-4815-ba16-6d98c0634efd)

Backtracing the function calls in the code, finding no palette files can have two main reasons
1. no files with `.gpl` extension in the palette directories found
    - The custom palettes is empty by default, so here it's expected to be empty :heavy_check_mark: 
    - The builtin palettes should get created during installation. 
    This should not be affected by removing the `settings.xml` file. Are you sure you didn't change anything else between the palettes not being there and the palettes reappearing?
2. the two [palette directories](https://github.com/idotobi/xournalpp/blob/13658feef3617a0382d420503a5f43145f5b408f/src/core/control/Control.cpp#L1225-L1226) pointing to the wrong directories
    - `Util::getConfigSubfolder("palettes")` cannot explain it since it's expected to be empty anyways :heavy_check_mark: 
    - `Util::getPalettePath()` on the other hand had an issue on MacOs that you found.Could you please double-check whether the fix you made in `PathUtil.cpp` indeed fixed this issue?

Sorry for asking so many questions again, but I'm running out of ideas for other causes to this issue.

idotobi commented:
@rolandlo regarding the Signal 6 crash during the toolbar customization:
I didn't really find a good explanation for why you still got a crash on your system, but at least I managed to cause a similar crash by changing the palette in the settings while being in the toolbar customization. This caused a crash because the toolbar customization dialog does currently not get updated when you change the settings (which is not ideal but I also would not expect this as normal user behaviour).
Prior to my last commit (https://github.com/xournalpp/xournalpp/pull/3767/commits/6acf5c6062e8aeb58bcefe6506d1b5f9ed35283b) pointers to `NamedColor`s where directly passed around which is dangerous if their lifetime ends for some reason before the user finishes with the toolbar customization. Since the `NamedColor` pointers pointed to the `NamedColor`s from the palette in the `Control` class (which lifes basically as long as the application runs) I didn't expext the lifetime to become an issue. Still passing pointers can easily result in a Signal 6 crash if some detail is overlooked.
Hence, I changed the code to rather pass the `paletteColorIndex` as a simple integer. This is sufficient for all usecases we currently support and should make it almost impossible to cause a Signal 6 crash.

Hopefully this finally fixes the issue :crossed_fingers:.

rolandlo commented:
> So lets look at it from a different direction. If I understood you correctly then you didn't see any palettes in the palette settings. So you probably see something like the following, right?
> 
> ![image](https://private-user-images.githubusercontent.com/16611056/327941478-2c38deba-310e-4815-ba16-6d98c0634efd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ4NDg1NTIsIm5iZiI6MTcxNDg0ODI1MiwicGF0aCI6Ii8xNjYxMTA1Ni8zMjc5NDE0NzgtMmMzOGRlYmEtMzEwZS00ODE1LWJhMTYtNmQ5OGMwNjM0ZWZkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTA0VDE4NDQxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTcyZjEwMDZkNmU0ZWI3OThjNGE1NzY5Y2YzYzNkZjhlZTE2NTUwZTE4MTEyNjA0NTUxMzQwNDRlZTQ1OGMwYzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.mF_tQh3qyzZ9bjnkkfzEZ5BQxFbTlQFgJoFp_bzz67Q)

Not quite. It looks like this:
![grafik](https://github.com/xournalpp/xournalpp/assets/40485433/fee6f904-7817-499f-bfaf-cafab5db155c)

And for sure, the only thing I changed was renaming (or deleting) the `settings.xml` to make it work again. And the only thing to make it not work was moving the `settings.xml` I uploaded above to `settings.xml` in the `.config/xournalpp` folder. It's mysterious, I know. Probably it's easier if I try to debug the problem. Maybe changing the settings.xml line by line.

rolandlo commented:
Looking at the code, it becomes clear: The difference is the disabled audio.

https://github.com/xournalpp/xournalpp/blob/13658feef3617a0382d420503a5f43145f5b408f/src/core/gui/dialog/SettingsDialog.cpp#L646-L661

Indeed when I enable audio it works again. Simply move `paletteTab.renderPaletteTab(this->control->getPalette().getFilePath());` out of the `    if (this->control->getAudioController()) { .... }` block.

idotobi commented:
🤦 Of course it's at the one place I didn't look closer as it seemed hard to f**k it up there. Thanks for looking into it! 🙏
I'll fix it right away.

rolandlo commented:
I'm glad it's solved now. I'll look into the toolbar customization later.

idotobi commented:
Sure, whenever you have time.
I'll work on a fix for the dark-mode update based on the OS setting in the meantime.

rolandlo commented:
Toolbar customization seems to work fine now as well. I didn't get any more crashes.

rolandlo commented:
For the issue with the secondary label colorization, I would simply remove all of 5c981c35999fb754784586a0abf4773ed55c40e9 except of the change:
```diff
-    GtkWidget* secondaryLabel = gtk_label_new(colorize(secondaryInformation, "gray").c_str());
+    GtkWidget* secondaryLabel = gtk_label_new(secondaryInformation.c_str());
```

idotobi commented:
I'll play around with getting the proper CSS information through a bit more, but if I don't find a god solution for this until next weekend I'll just fall back to the default string color as you suggested :+1: 

The last commit almost did it, but the signal handler needs to be cleaned up which is not that easy as the `PopupWindowWrapper` is destructured before the window closes :sweat_smile:

idotobi commented:
> For the issue with the secondary label colorization, I would simply remove all of [5c981c3](https://github.com/xournalpp/xournalpp/commit/5c981c35999fb754784586a0abf4773ed55c40e9) except of the change:
> 
> ```diff
> -    GtkWidget* secondaryLabel = gtk_label_new(colorize(secondaryInformation, "gray").c_str());
> +    GtkWidget* secondaryLabel = gtk_label_new(secondaryInformation.c_str());
> ```

@rolandlo : I ended up following your sugggestion. The `PopupWindowWrapper` didn't really lend itself to properly implement the css propagation I had in mind, and it's not really worth it to further postpone this feature just for that subfeature.

idotobi commented:
From my side it should be ready for merging now. 
If you agree we can ask Febbe to check if it's also properly running on Windows.

Regarding the merge: I would rebase and merge it once both of you give the OK. 
Do you have an opinion on whether I should _(A)_ squash the commits down to one commit, _(B)_ squash the commits down to some bigger work packages (i.e. ~5-10 commits), or _(C)_ keep all of the 45 commits? 
I'd probably prefer (B).

rolandlo commented:
> Regarding the merge: I would rebase and merge it once both of you give the OK. Do you have an opinion on whether I should _(A)_ squash the commits down to one commit, _(B)_ squash the commits down to some bigger work packages (i.e. ~5-10 commits), or _(C)_ keep all of the 45 commits? I'd probably prefer (B).

My opinion is clearly (B).

rolandlo commented:
I have tested the PR brievely on Windows. After I had upgraded MSYS2, it didn't compile for me. Including the `<algorithm>` header in `PathUtil.cpp` (for `std::sort`) fixed that. After that it worked normally on Windows, except that the preferences window default size is a bit too small to see the palettes on the very right without scrolling or enlarging the window.
The preferences window is too small on windows for other reasons as well, so maybe we should just enlarge it a little by default.

idotobi commented on src/core/gui/dialog/SettingsDialog.cpp:
```diff
@@ -1038,6 +1048,10 @@ void SettingsDialog::save() {
     settings->transactionEnd();
 
     this->control->getWindow()->setGtkTouchscreenScrollingForDeviceMapping();
+    this->control->loadPaletteFromSettings();
+    this->control->getWindow()->getToolMenuHandler()->updateColorToolItems(this->control->getPalette());
+    this->control->getWindow()->reloadToolbars();
+
```
I'll also change the name of `callbackData` to `settingsBeforeDialog` in the `Control::showSettings()` to make the logic a bit more obvious.

idotobi commented on src/core/gui/dialog/SettingsDialogPaletteTab.cpp:
```diff
@@ -0,0 +1,200 @@
+    gtk_box_pack_start(GTK_BOX(rowContent), text, false, false, 0);
+
+    GtkWidget* colorIcons = newPaletteColorIconsBox(palette);
+    gtk_box_pack_start(GTK_BOX(rowContent), colorIcons, true, true, 0);
```
Done :heavy_check_mark: 
The only istance of a function mentioned in https://github.com/xournalpp/xournalpp/issues/2722#issuecomment-1925438910 was `gtk_widget_show_all` which I assume is fine to use.

idotobi commented:
@rolandlo:
I added the `<algorithm>` dependency to `PathUtil.cpp` (https://github.com/idotobi/xournalpp/commit/89d87193193096d3faf7dc24029d94321fb2efff).
Furthermore, I changed the UI a bit to make it more adaptable to smaller dialog windows, even for longer palette paths (https://github.com/idotobi/xournalpp/commit/646a3fd9a2680245365e8d49ad89bb2cb5b932eb):
![image](https://github.com/xournalpp/xournalpp/assets/16611056/efa54159-eaa7-4495-9cfb-cc5749bea923)

rolandlo commented on src/core/gui/toolbarMenubar/ToolMenuHandler.cpp:
```diff
@@ -98,7 +98,8 @@ void ToolMenuHandler::unloadToolbar(GtkWidget* toolbar) {
 
-    const Palette& palette = this->control->getSettings()->getColorPalette();
+    const auto palette = this->control->getPalette();
+    size_t colorIndex{};
```
This variable is unused.
```suggestion
```

rolandlo reviewed (approved):
It works all fine now. The settings dialog is large enough to display the palettes on Windows.
Please remove the unused variable `colorIndex` though.

idotobi commented on src/core/gui/dialog/SettingsDialog.h:
```diff
@@ -26,20 +26,23 @@
     SettingsDialog(GladeSearchpath* gladeSearchPath, Settings* settings, Control* control,
-                   std::function<void()> callback);
+                   const std::vector<fs::path>& paletteDirectories, std::function<void()> callback);
 
     inline GtkWindow* getWindow() const { return window.get(); }
-
-private:
     void save();
 
+private:
```
No, it only was needed for a test which I had to remove again, because it was to unstable.

idotobi commented on src/core/control/XournalMain.h:
```diff
@@ -12,8 +12,13 @@
 auto run(int argc, char** argv) -> int;
-}
+
+void initResourcePath(GladeSearchpath* gladePath, std::string_view relativePathAndFile, bool failIfNotFound = true);
```
True, now that the test is removed again these changes are not really needed.

idotobi commented:
@bhennion : Thanks for the detailed review. I fixed your findings. Please have a look again :slightly_smiling_face:

idotobi commented:
@bhennion : I rewrote the branch history as you suggested and kept the original one at https://github.com/idotobi/xournalpp/tree/backup/palette-settings. The only additional change I did is that I removed empty lines that were added for no reason to `XournalMain.h` in the original branch.
Is this what you had in mind?

KronosTheLate commented:
Thanks for the hard work to everyone who contributed <3

Mercadinho1 commented:
heyy, wassup! Can you explain to me how I can install this "mod" on my
xournal? I don't know anything about programming, but I was following the
process to install later...

Em ter., 23 de jul. de 2024 07:06, KronosTheLate ***@***.***>
escreveu:

> Thanks for the hard work to everyone who contributed <3
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/xournalpp/xournalpp/pull/3767#issuecomment-2244792537>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AURWKU5IEBM64MCXACL5NSTZNYTMBAVCNFSM5NP6EXAKU5DIOJSWCZC7NNSXTN2JONZXKZKDN5WW2ZLOOQ5TEMRUGQ3TSMRVGM3Q>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

KronosTheLate commented:
> heyy, wassup! Can you explain to me how I can install this "mod" on my
> xournal? I don't know anything about programming, but I was following the
> process to install later...
> 
> Em ter., 23 de jul. de 2024 07:06, KronosTheLate ***@***.***>
> escreveu:
> 
> > Thanks for the hard work to everyone who contributed <3
> >
> > —
> > Reply to this email directly, view it on GitHub
> > <https://github.com/xournalpp/xournalpp/pull/3767#issuecomment-2244792537>,
> > or unsubscribe
> > <https://github.com/notifications/unsubscribe-auth/AURWKU5IEBM64MCXACL5NSTZNYTMBAVCNFSM5NP6EXAKU5DIOJSWCZC7NNSXTN2JONZXKZKDN5WW2ZLOOQ5TEMRUGQ3TSMRVGM3Q>
> > .
> > You are receiving this because you are subscribed to this thread.Message
> > ID: ***@***.***>
> >
> 

I guess you have to clone the master branch og the repository locally, and then build Xournal locally. If you are new to cloning repositories and building locally, and do not feel adventurous, i reccomend wayting until this feature makes it into a tagged release.
